src/dev/src/app/datagrid/utils/color-filter.ts

```diff
@@ -9,42 +9,49 @@ import { User } from '../inventory/user';
 import { COLORS } from '../inventory/values';
 
 @Component({
-  selector: 'clr-datagrid-color-filter-demo',
-  template: `
+    selector: 'clr-datagrid-color-filter-demo',
+    template: `
         <span *ngFor="let color of allColors" class="color-square color-selectable"
             (click)="toggleColor(color)" 
             [style.backgroundColor]="color"
             [class.color-selected]="selectedColors[color]"></span>`,
-  styleUrls: ['../datagrid.demo.scss'],
+    styleUrls: ['../datagrid.demo.scss'],
 })
 export class ColorFilter implements ClrDatagridFilterInterface<User> {
-  allColors = COLORS;
-  selectedColors: { [color: string]: boolean } = {};
-  nbColors = 0;
+    allColors = COLORS;
+    selectedColors: { [color: string]: boolean } = {};
+    nbColors = 0;
 
-  changes: EventEmitter<any> = new EventEmitter<any>(false);
+    changes: EventEmitter<any> = new EventEmitter<any>(false);
 
-  listSelected(): string[] {
-    const list: string[] = [];
-    for (const color in this.selectedColors) {
-      if (this.selectedColors[color]) {
-        list.push(color);
-      }
+    listSelected(): string[] {
+        const list: string[] = [];
+        for (const color in this.selectedColors) {
+            if (this.selectedColors[color]) {
+                list.push(color);
+            }
+        }
+        return list;
     }
-    return list;
-  }
 
-  toggleColor(color: string) {
-    this.selectedColors[color] = !this.selectedColors[color];
-    this.selectedColors[color] ? this.nbColors++ : this.nbColors--;
-    this.changes.emit(true);
-  }
+    toggleColor(color: string) {
+        this.selectedColors[color] = !this.selectedColors[color];
+        this.selectedColors[color] ? this.nbColors++ : this.nbColors--;
+        this.changes.emit(true);
+    }
 
-  accepts(user: User) {
-    return this.nbColors === 0 || this.selectedColors[user.color];
-  }
+    accepts(user: User) {
+        return this.nbColors === 0 || this.selectedColors[user.color];
+    }
 
-  isActive(): boolean {
-    return this.nbColors > 0;
-  }
+    isActive(): boolean {
+        return this.nbColors > 0;
+    }
+
+    /**
+       * Compare objects by reference
+       */
+    public equals(objectToCompare: ColorFilter): boolean {
```

Comment:
This problematic and shows the main difficulty with this approach. The filter object that is passed in the input state must be the _same_ as the registered one, otherwise a new filter is added. You therefore cannot pass a differently configured filter in the input state, because the only the way to do this would be to modify the existing filter, which already changes the state.

For this to work, filters must be serializable or have a state input and output.

Reply:
For the string filter the equals is ok to check prop name? For color filters may also need an identifier prop name, to determine which filter uses it, and with that we could implement the equals. Or am I wrong?

Reply:
The string filters also needs to compare the value and, if the value is not the same, you must set the value in the string filter to activate or deactivate it. Basically, you need to have getters and setters for the internal state of the filter.